### PR TITLE
Go reportcard fix

### PR DESCRIPTION
**Pull request by last-saiyan**
improved `golint` coverage to 60%   [issue](https://github.com/juicedata/juicefs/issues/77)

to test the coverage setup `goreportcard` on local machine and  run `goreportcard-cli`

### CODE REVIEW

File: pkg/utils/alloc.go [hunk 1]
```diff
@@ -26,6 +26,10 @@ var slabs = make(map[uintptr][]byte)
 var used int64
 var slabsMutex sync.Mutex
 
+// Alloc allocates memory for page returns slice of byte of lenght `size`
```
Comment (davies) [inline pkg/utils/alloc.go:29]: lenght -> length

File: pkg/utils/alloc.go [hunk 2]
```diff
@@ -26,6 +26,10 @@ var slabs = make(map[uintptr][]byte)
...
+// adds page entry in slabs map
+// adds `size` `used` memory
```
Comment (davies) [inline pkg/utils/alloc.go:31]: We don't need to add these details in comments

File: pkg/utils/cond.go [hunk 3]
```diff
@@ -54,6 +58,8 @@ var timerPool = sync.Pool{
 	},
 }
 
+// WaitWithTimeout waits till timeout or message in signal channel
+// returns true incase of timeout else false
```
Comment (davies) [inline pkg/utils/cond.go:62]: incase -> in case

File: pkg/utils/cond.go [hunk 4]
```diff
@@ -20,18 +20,21 @@ import (
 	"time"
 )
 
+// Cond struct contains a channel and lock used for synchronization
```
Comment (davies) [inline pkg/utils/cond.go:23]: Cond is similar to sync.Cond, but you can wait without a timeout.

File: pkg/utils/cond.go [hunk 5]
```diff
@@ -20,18 +20,21 @@ import (
...
 
+// Signal sends true to the channel
 func (c *Cond) Signal() {
 	select {
 	case c.signal <- true:
 	default:
 	}
 }
 
+// Broadcast sends true to channel
```
Comment (davies) [inline pkg/utils/cond.go:37]: Broadcast wake up all the waiters.

File: pkg/utils/cond.go [hunk 6]
```diff
@@ -20,18 +20,21 @@ import (
 	"time"
 )
 
+// Cond struct contains a channel and lock used for synchronization
 type Cond struct {
 	L      sync.Locker
 	signal chan bool
 }
 
+// Signal sends true to the channel
```
Comment (davies) [inline pkg/utils/cond.go:29]: Signal wakes up a waiter.

File: pkg/utils/cond.go [hunk 7]
```diff
@@ -42,6 +45,7 @@ func (c *Cond) Broadcast() {
 	}
 }
 
+// Wait till signal is recieived
```
Comment (davies) [inline pkg/utils/cond.go:48]: Wait until Signal() or Broadcast() is called.

File: pkg/utils/cond.go [hunk 8]
```diff
@@ -54,6 +58,8 @@ var timerPool = sync.Pool{
 	},
 }
 
+// WaitWithTimeout waits till timeout or message in signal channel
```
Comment (davies) [inline pkg/utils/cond.go:61]: WaitWithTimeout wait for a signal or a period of `timeout` eclipsed.

File: pkg/utils/cond.go [hunk 9]
```diff
@@ -71,6 +77,7 @@ func (c *Cond) WaitWithTimeout(d time.Duration) bool {
 	}
 }
 
+// NewCond creates & returns pointer to new Cond
```
Comment (davies) [inline pkg/utils/cond.go:80]: NewCond creates a Cond.

File: pkg/utils/alloc.go [hunk 10]
```diff
@@ -36,6 +40,9 @@ func Alloc(size int) []byte {
 	return b
 }
 
+// Free removes entry of page from slabs map
+// removes `size` from `used` memory
+// function is thread safe
```
Comment (davies) [inline pkg/utils/alloc.go:45]: Free returns memory to Go heap.

File: pkg/utils/buffer.go [hunk 11]
```diff
@@ -20,111 +20,134 @@ import (
 	"unsafe"
 )
 
+// Buffer struct containing slice of byte
```
Comment (davies) [inline pkg/utils/buffer.go:23]: Buffer is a buffer to read/write integers.

File: pkg/utils/buffer.go [hunk 12]
```diff
@@ -20,111 +20,134 @@ import (
...
+// it has methods append and remove bytes based on datatype and byte ordring
 type Buffer struct {
 	endian binary.ByteOrder
 	off    int
 	buf    []byte
 }
 
+// NewBuffer utility to create *Buffer of given size
```
Comment (davies) [inline pkg/utils/buffer.go:31]: NewBuffer returns a buffer with `sz` number of bytes.

**Review by davies — CHANGES_REQUESTED** [review]
To have a better doc for a function, we should write `the thing it can archive`, not how it is implemented.

**Comment by suzaku** [comment]
> To have a better doc for a function, we should write `the thing it can archive`, not how it is implemented.

Maybe these kind of doc are better written by the authors 🤦 Other people have to read the code to understand them and tend to write about the implementation.

**Pull request merged**